saphir/src/http_context.rs

```diff
@@ -1,19 +1,412 @@
-use crate::{request::Request, router::Router};
+use crate::{request::Request, response::Response, router::Router};
+
+/// State of the Http context. It represent whether the context is used
+/// `Before(..)` or `After(..)` calling the handler responsible of generating a
+/// responder. Empty will be the state of a context when the request is being
+/// processed by the handler, or when its original state has been moved by using
+/// take & take unchecked methods
+pub enum State {
+    Before(Request),
+    After(Response),
+    Empty,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        State::Empty
+    }
+}
+
+impl State {
+    /// Take the current context leaving `State::Empty` behind
+    pub fn take(&mut self) -> Self {
+        std::mem::take(self)
+    }
+
+    /// Take the current request leaving `State::Empty` behind
+    /// Returns `Some(Request)` if the state was `Before` or `None` if it was
+    /// something else
+    pub fn take_request(&mut self) -> Option<Request> {
+        match std::mem::take(self) {
+            State::Before(r) => Some(r),
+            _ => None,
+        }
+    }
+
+    /// Take the current request leaving `State::Empty` behind
+    /// Panics if the state is not `Before`
```

Comment:
per the rust api doc guideline, the panics line should be under the panics section like this `/// # Panics`
see https://rust-lang.github.io/api-guidelines/documentation.html

Reply:
done